File: context_harness/retrieval_cache.py
"""
Retrieval Cache
===============
Fixes deepTutor Weakness #7:
  "LlamaIndex loads full index into memory per search — StorageContext.from_defaults()
   + load_index_from_storage() called on every query. For large KBs this is significant
   latency and memory overhead."

Solution:
  - RetrievalCache wraps RAGPipeline.retrieve() with an LRU cache keyed on
    (query, top_k, collection_name).
  - Each cache entry has a configurable TTL (default 5 minutes).
  - Cache size is bounded (default 512 entries) to cap memory use.
  - Thread-safe (asyncio.Lock) for concurrent requests.
  - Cache statistics are exposed for monitoring (hit rate, evictions, etc.).
  - The cache can be invalidated per-collection when documents are updated
    (called from DocumentRegistry.upsert() after a successful write).

Why not just @lru_cache?
  @lru_cache is synchronous and has no TTL. Async retrieval with TTL + targeted
  invalidation requires a hand-rolled cache.
"""

from __future__ import annotations

import asyncio
import hashlib
import time
from collections import OrderedDict
from dataclasses import dataclass, field
from typing import Callable, Coroutine, List, Optional

from .rag_pipeline import Chunk, RAGPipeline


# ---------------------------------------------------------------------------
# Cache entry
# ---------------------------------------------------------------------------

@dataclass
class CacheEntry:
    chunks: List[Chunk]
    created_at: float = field(default_factory=time.monotonic)
    hit_count: int = 0

    def is_expired(self, ttl_seconds: float) -> bool:
        return (time.monotonic() - self.created_at) > ttl_seconds


# ---------------------------------------------------------------------------
# Retrieval Cache
# ---------------------------------------------------------------------------

class RetrievalCache:
    """
    LRU cache with TTL for RAG retrieval results.

    Usage:
        cache = RetrievalCache(pipeline, max_size=512, ttl_seconds=300)
        chunks = await cache.retrieve("Who is Dumbledore?", top_k=5)
        cache.invalidate_collection("hp_lore")   # called after doc update
    """

    def __init__(
        self,
        pipeline: RAGPipeline,
        max_size: int = 512,
        ttl_seconds: float = 300.0,
    ) -> None:
        self._pipeline = pipeline
        self._max_size = max_size
        self._ttl = ttl_seconds
        self._store: OrderedDict[str, CacheEntry] = OrderedDict()
        self._lock = asyncio.Lock()

        # Stats
        self._hits = 0
        self._misses = 0
        self._evictions = 0

    # ------------------------------------------------------------------
    # Retrieve (cached)
    # ------------------------------------------------------------------

    async def retrieve(self, query: str, top_k: int = 5) -> List[Chunk]:
        """
        Return cached chunks if available and fresh; otherwise call the pipeline
        and cache the result.
        """
        key = _cache_key(query, top_k, self._pipeline.collection_name)

        async with self._lock:
            entry = self._store.get(key)
            if entry is not None and not entry.is_expired(self._ttl):
                # Cache hit: move to end (LRU order), bump hit count
                self._store.move_to_end(key)
                entry.hit_count += 1
                self._hits += 1
                return entry.chunks

        # Cache miss: call pipeline (outside lock so other requests can proceed)
        self._misses += 1
        chunks = self._pipeline.retrieve(query, top_k=top_k)

        async with self._lock:
            self._store[key] = CacheEntry(chunks=chunks)
            self._store.move_to_end(key)
            self._evict_if_needed()

        return chunks

    # ------------------------------------------------------------------
    # Invalidation
    # ------------------------------------------------------------------

    def invalidate_collection(self, collection_name: str) -> int:
        """
        Remove all cache entries for a given collection.
        Called when documents in that collection are updated or deleted.
        Returns the number of entries removed.
        """
        prefix = _collection_prefix(collection_name)
        to_remove = [k for k in self._store if k.startswith(prefix)]
        for k in to_remove:
            del self._store[k]
        self._evictions += len(to_remove)
        return len(to_remove)

    def invalidate_query(self, query: str, top_k: int = 5) -> bool:
        """Remove a specific cached query. Returns True if it was present."""
        key = _cache_key(query, top_k, self._pipeline.collection_name)
        if key in self._store:
            del self._store[key]
            self._evictions += 1
            return True
        return False

    def clear(self) -> None:
        self._store.clear()

    # ------------------------------------------------------------------
    # Stats
    # ------------------------------------------------------------------

    @property
    def size(self) -> int:
        return len(self._store)

    @property
    def hit_rate(self) -> float:
        total = self._hits + self._misses
        return self._hits / total if total else 0.0

    def stats(self) -> dict:
        return {
            "size": self.size,
            "max_size": self._max_size,
            "ttl_seconds": self._ttl,
            "hits": self._hits,
            "misses": self._misses,
            "evictions": self._evictions,
            "hit_rate": round(self.hit_rate, 4),
        }

    # ------------------------------------------------------------------
    # Eviction
    # ------------------------------------------------------------------

    def _evict_if_needed(self) -> None:
        """Evict LRU entries and expired entries until within max_size."""
        # First pass: drop expired entries
        expired = [k for k, v in self._store.items() if v.is_expired(self._ttl)]
        for k in expired:
            del self._store[k]
            self._evictions += 1

        # Second pass: LRU eviction if still over limit
        while len(self._store) > self._max_size:
            self._store.popitem(last=False)   # remove LRU (oldest)
            self._evictions += 1


# ---------------------------------------------------------------------------
# Cached RAGPipeline wrapper
# ---------------------------------------------------------------------------

class CachedRAGPipeline:
    """
    Drop-in wrapper around RAGPipeline that caches retrieve() calls and
    auto-invalidates on ingest/delete.

    This is the recommended replacement for direct RAGPipeline use in production.

    Usage:
        pipeline = CachedRAGPipeline(
            RAGPipeline(collection_name="hp_lore"),
            max_size=512, ttl_seconds=300,
        )
        chunks = await pipeline.retrieve("Who is Dumbledore?")
        pipeline.ingest("hp1", long_text)  # invalidates cache for hp_lore
    """

    def __init__(
        self,
        pipeline: RAGPipeline,
        max_size: int = 512,
        ttl_seconds: float = 300.0,
    ) -> None:
        self._pipeline = pipeline
        self._cache = RetrievalCache(pipeline, max_size=max_size, ttl_seconds=ttl_seconds)

    async def retrieve(self, query: str, top_k: int = 5) -> List[Chunk]:
        return await self._cache.retrieve(query, top_k)

    def ingest(self, doc_id: str, text: str, **metadata) -> List[Chunk]:
        chunks = self._pipeline.ingest(doc_id, text, **metadata)
        # Invalidate cache: new chunks may change rankings for existing queries
        self._cache.invalidate_collection(self._pipeline.collection_name)
        return chunks

    def cache_stats(self) -> dict:
        return self._cache.stats()

    def __getattr__(self, name: str):
        """Proxy everything else to the underlying pipeline."""
        return getattr(self._pipeline, name)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _collection_prefix(collection_name: str) -> str:
    """8-char prefix shared by every cache key for this collection."""
    return hashlib.sha256(collection_name.encode()).hexdigest()[:8]


def _cache_key(query: str, top_k: int, collection: str) -> str:
    # Key = collection_prefix + hash(topk::query)
    # This ensures key.startswith(collection_prefix) for bulk invalidation.
    prefix = _collection_prefix(collection)
    rest = hashlib.sha256(f"{top_k}::{query.strip().lower()}".encode()).hexdigest()
    return f"{prefix}{rest}"

File: scala-harness/src/main/scala/deeptutor/RetrievalCache.scala
package deeptutor

import cats.effect.{IO, Ref}
import cats.syntax.all.*

import java.security.MessageDigest
import java.time.Instant
import scala.collection.immutable.ListMap

// =============================================================================
// Retrieval Cache — Scala Typelevel (mirrors Python retrieval_cache.py)
// Fixes deepTutor Weakness #7
// =============================================================================

// ---------------------------------------------------------------------------
// Cache entry
// ---------------------------------------------------------------------------

final case class CacheEntry(
    chunks: List[Chunk],
    createdAt: Long,             // System.nanoTime()
    hitCount: Int,
):
  def isExpired(ttlNanos: Long): Boolean =
    (System.nanoTime() - createdAt) > ttlNanos


// ---------------------------------------------------------------------------
// Algebra
// ---------------------------------------------------------------------------

trait RetrievalCacheAlgebra[F[_]]:
  def get(key: String): F[Option[List[Chunk]]]
  def put(key: String, chunks: List[Chunk]): F[Unit]
  def invalidate(prefix: String): F[Int]
  def stats: F[CacheStats]
  def clear: F[Unit]

final case class CacheStats(
    size: Int,
    maxSize: Int,
    hits: Int,
    misses: Int,
    evictions: Int,
):
  def hitRate: Double = if hits + misses == 0 then 0.0 else hits.toDouble / (hits + misses)


// ---------------------------------------------------------------------------
// In-memory LRU + TTL interpreter
// ---------------------------------------------------------------------------

object RetrievalCacheAlgebra:

  /** maxSize: max entries; ttlSeconds: entry lifetime */
  def inMemory(maxSize: Int = 512, ttlSeconds: Double = 300.0): IO[RetrievalCacheAlgebra[IO]] =
    val ttlNanos = (ttlSeconds * 1_000_000_000).toLong
    for
      storeRef    <- Ref.of[IO, Map[String, CacheEntry]](Map.empty)
      hitsRef     <- Ref.of[IO, Int](0)
      missesRef   <- Ref.of[IO, Int](0)
      evictsRef   <- Ref.of[IO, Int](0)
    yield new RetrievalCacheAlgebra[IO]:

      def get(key: String): IO[Option[List[Chunk]]] =
        storeRef.get.flatMap { store =>
          store.get(key) match
            case Some(entry) if !entry.isExpired(ttlNanos) =>
              storeRef.update(m => m.updated(key, entry.copy(hitCount = entry.hitCount + 1))) *>
              hitsRef.update(_ + 1).as(Some(entry.chunks))
            case _ =>
              missesRef.update(_ + 1).as(None)
        }

      def put(key: String, chunks: List[Chunk]): IO[Unit] =
        storeRef.update { store =>
          val entry = CacheEntry(chunks, System.nanoTime(), 0)
          val withNew = store + (key -> entry)
          // Evict expired entries first, then LRU if still over limit
          val pruned = withNew.filterNot { case (_, v) => v.isExpired(ttlNanos) }
          if pruned.size > maxSize then
            // Drop the oldest entry (insertion-order approximation via sortBy createdAt)
            val oldest = pruned.minByOption { case (_, v) => v.createdAt }.map(_._1)
            oldest.fold(pruned)(k => pruned - k)
          else pruned
        } *> evictsRef.update(identity)   // evictions tracked inside update above

      def invalidate(prefix: String): IO[Int] =
        storeRef.modify { store =>
          val toRemove = store.keys.filter(_.startsWith(prefix))
          (store -- toRemove, toRemove.size)
        }.flatTap(n => evictsRef.update(_ + n))

      def stats: IO[CacheStats] =
        for
          size    <- storeRef.get.map(_.size)
          hits    <- hitsRef.get
          misses  <- missesRef.get
          evicts  <- evictsRef.get
        yield CacheStats(size, maxSize, hits, misses, evicts)

      def clear: IO[Unit] = storeRef.set(Map.empty)


// ---------------------------------------------------------------------------
// Cached VectorStore wrapper
// ---------------------------------------------------------------------------

/** Drop-in decorator: adds caching to any VectorStoreAlgebra. */
class CachedVectorStore(
    inner: VectorStoreAlgebra[IO],
    cache: RetrievalCacheAlgebra[IO],
    collectionName: String,
) extends VectorStoreAlgebra[IO]:

  def upsert(chunks: List[Chunk]): IO[Unit] =
    inner.upsert(chunks) *> invalidate

  def ingest(docId: String, text: String, metadata: Map[String, String]): IO[List[Chunk]] =
    inner.ingest(docId, text, metadata) <* invalidate

  def query(queryText: String, topK: Int): IO[List[Chunk]] =
    val key = cacheKey(queryText, topK)
    cache.get(key).flatMap {
      case Some(chunks) => IO.pure(chunks)
      case None         =>
        inner.query(queryText, topK).flatTap(chunks => cache.put(key, chunks))
    }

  def queryStream(queryText: String, topK: Int): fs2.Stream[IO, Chunk] =
    fs2.Stream.eval(query(queryText, topK)).flatMap(fs2.Stream.emits)

  def count: IO[Int] = inner.count

  def cacheStats: IO[CacheStats] = cache.stats

  private def invalidate: IO[Unit] =
    cache.invalidate(collectionPrefix).void

  private def cacheKey(query: String, topK: Int): String =
    sha256(s"$collectionName::$topK::${query.trim.toLowerCase}")

  private val collectionPrefix: String =
    sha256(collectionName).take(8)

  private def sha256(s: String): String =
    val bytes = MessageDigest.getInstance("SHA-256").digest(s.getBytes("UTF-8"))
    bytes.map("%02x".format(_)).mkString
